battle.go

```diff
@@ -55,214 +23,128 @@ const (
+func (b *Battle) SimulateRound() {
+	// Collects all turn info from each active Pokemon
+	turns := make([]TurnContext, 0)
+	for _, party := range b.parties {
+		for _, pokemon := range party.activePokemon {
+			ctx := b.GetContext(party, pokemon)
+			turn := (*party.Agent).Act(ctx)
+			turns = append(turns, TurnContext{Turn: &turn, Context: ctx})
 		}
-	} else {
-		// make higher priority turns go first
-		return tA.Priority() > tB.Priority()
 	}
-	// fallthrough
-	return false
-}
-
-// Returns the indexes of the active pokemon in the order that their turns should take place.
-func sortTurns(battle *Battle, ap []activePokemon, turns map[int]Turn) []int {
-	t := newTurnOrder(battle, ap, turns)
-	sort.Sort(&t)
-	return t.GetOrder()
-}
-
-// References a Pokemon currently on the battlefield.
-type activePokemon struct {
-	PartyIdx   int
-	PokemonIdx int
-}
-
-// Used to provide data to Agents such that the Agent can't maliciously modify the state of the battle.
-type BattleInfo interface {
-	GetPokemon(int) Pokemon
-	Self() int
-	Allies() []int
-	Opponents() []int
-}
-
-// Implements the BattleInfo interface. Passed to Agents so they can decide a turn.
-type battleContext struct {
-	Context       Battle
-	ActivePokemon []activePokemon
-	self          int
-}
-
-func (c *battleContext) getTeam() []int {
-	selfParty := c.ActivePokemon[c.self].PartyIdx
-	var team []int
-	for _, t := range c.Context.teams {
-		if contains(t, selfParty) {
-			team = t
-			break
+	// Sort turns using an in-place stable sort
+	sort.SliceStable(turns, func(i, j int) bool {
+		turnA := turns[i].Turn
+		turnB := turns[j].Turn
+		ctxA := turns[i].Context
+		ctxB := turns[j].Context
+		if reflect.TypeOf(*turnA) == reflect.TypeOf(*turnB) {
+			switch (*turnA).(type) {
+			case FightTurn:
+				// speedy pokemon should go first
+				return ctxA.Pokemon.Stats[STAT_SPD] > ctxB.Pokemon.Stats[STAT_SPD]
+			}
+		} else {
+			// make higher priority turns go first
+			return (*turnA).Priority() > (*turnB).Priority()
 		}
-	}
-	return team
-}
-func (c *battleContext) GetPokemon(idx int) Pokemon {
-	p := c.Context.derefActivePokemon(c.ActivePokemon[idx])
-	return *p
-}
-func (c *battleContext) SetSelf(idx int) {
-	c.self = idx
-}
-
-// Get the index of the active Pokemon that this battle context is referencing.
-// In other words, the Pokemon you are acting as.
-func (c *battleContext) Self() int {
-	return c.self
-}
-func (c *battleContext) Allies() []int {
-	team := c.getTeam()
-
-	allies := []int{}
-	for i, ap := range c.ActivePokemon {
-		if i != c.self && contains(team, ap.PartyIdx) {
-			allies = append(allies, i)
+		// fallthrough
+		return false
+	})
+	// Run turns in sorted order and update battle state
+	for _, turn := range turns {
+		switch t := (*turn.Turn).(type) {
+		case FightTurn:
+			log.Printf("TODO: Implement fight\n")
+		default:
+			log.Panicf("Unknown turn of type %v", t)
 		}
 	}
+}
 
-	return allies
+type Context struct {
```

Comment:
(nit) to avoid name conflicts with other packages
```suggestion
type BattleContext struct {
```

Does this struct need to be public?

Reply:
Yes, this type is a parameter in `agent.Act(ctx)`.